ir/instr.cpp

```diff
@@ -1687,19 +1687,30 @@ static void unpack_inputs(State &s, Value &argv, Type &ty,
 
   auto unpack = [&](StateValue &&value) {
     if (ty.isPtrType()) {
+      expr np = value.non_poison;
+
       Pointer p(s.getMemory(), move(value.value));
       p.stripAttrs();
       if (argflag.has(ParamAttrs::Dereferenceable) ||
           argflag.has(ParamAttrs::ByVal))
         s.addUB(
           p.isDereferenceable(argflag.getDerefBytes(), argflag.align, false));
-      else if (argflag.has(ParamAttrs::Align))
-        s.addUB(p.isAligned(argflag.align));
+      else if (argflag.has(ParamAttrs::Align)) {
+        // align and nonnull: the arg becomes poison if the condition isn't met
+        if (argflag.poisonImpliesUB())
```

Comment:
instead of doing all this case splitting on poisonImpliesUB() which leads to a lot of code duplication, I suggest to collect all poison first and then move it to UB().